File: other/past202005-open/J.py
import sys
from bisect import bisect_left

sys.setrecursionlimit(10 ** 7)
rl = sys.stdin.readline


def solve():
    N, M = map(int, rl().split())
    a = list(map(int, rl().split()))
    
    ans = [-1] * M
    highest = [0] * N
    for i in range(M):
        idx = bisect_left(highest, a[i]) - 1
        if idx != -1:
            highest[idx] = a[i]
            ans[i] = N - idx
    print(*ans, sep='\n')


if __name__ == '__main__':
    solve()

File: other/past202005-open/J.nim
import algorithm, sequtils, strutils


var a, highest, ans: seq[int]


proc solve() =
    var N, M: int
    (N, M) = stdin.readLine.split.map(parseInt)
    a = stdin.readLine.split.map(parseInt)

    highest = newSeq[int](N)
    ans = newSeqWith(M, -1)
    var idx: int
    for i, ai in a:
        idx = highest.lower_bound(ai) - 1
        if idx != -1:
            highest[idx] = ai
            ans[i] = N - idx
    echo ans.join("\n")


when is_main_module:
    solve()
